### PR TITLE
Extend LEAD/LAG to work with non-fixed-width types

### PR DESCRIPTION
**Pull request by mythrocks**
This commit extends the [LEAD()/LAG() offset window functions](https://learnsql.com/blog/lead-and-lag-functions-in-sql/) introduced in #6277 to work with more than just fixed-width data types. The functions are defined as follows:

1. `LEAD(N)`: Returns the row from the input column, at the specified offset past the
current row. If the offset crosses the grouping boundary or column boundary for
a given row, a "default" value is returned if available. The "default" value is
null, by default.
1. `LAG(N)`: returns the row from the input column at the specified offset preceding
the current row. If the offset crosses the grouping boundary or column boundary for
a given row, a "default" value is returned if available. The "default" value is
null, by default.

As an illustration, consider the following example input array input column, with two groups (`G1` and `G2`):
```c++
[ [1,1,1], [2], [3,3,3], [4,4,4,4],  [66,66], [], [88,88], [99] ]
  <--------------G1-------------->   <------------G2----------->
```

`LEAD(col, 1)` yields:
```c++
[ [2], [3,3,3], [4,4,4,4], ∅, [], [88,88], [99], ∅ ]
```

`LAG(input_col, 2)` yields:
```c++
[ ∅, ∅, [2], [3,3,3], ∅, ∅, [], [88,88] ]
```

If a `defaults` column is specified with contents:
```c++
[ [999], [999], [999], [999], [999], [999], [999], [999] ]
```
then, `LEAD(col, 1, defaults)` yields:
```c++
[ [2], [3,3,3], [4,4,4,4], [999], [], [88,88], [99], [999] ]
```
Note that in the cases where the offset (`1`) would cross the column/group boundary (i.e. indices `3` and `7`), the corresponding entry from the `defaults` column is returned instead of `∅` (i.e. `null`).

### CODE REVIEW

**Comment by mythrocks** [comment]
Looks to be a transient test failure:
```
11:21:25 FATAL: Non-zero status 255: [nvidia-smi, -L]
11:21:25 java.lang.RuntimeException: Non-zero status 255: [nvidia-smi, -L]
11:21:25 	at com.gpuopenanalytics.jenkins.remotedocker.config.NvidiaGpuDevicesConfigItem.executeWithOutput(NvidiaGpuDevicesConfigItem.java:130)
11:21:25 	at com.gpuopenanalytics.jenkins.remotedocker.config.NvidiaGpuDevicesConfigItem.addCreateArgs(NvidiaGpuDevicesConfigItem.java:83)
11:21:25 	at com.gpuopenanalytics.jenkins.remotedocker.job.DockerImageConfiguration.lambda$addCreateArgs$0(DockerImageConfiguration.java:140)
11:21:25 	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384)
11:21:25 	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:647)
11:21:25 	at com.gpuopenanalytics.jenkins.remotedocker.job.DockerImageConfiguration.addCreateArgs(DockerImageConfiguration.java:140)
11:21:25 	at com.gpuopenanalytics.jenkins.remotedocker.DockerState.getlaunchArgs(DockerState.java:275)
11:21:25 	at com.gpuopenanalytics.jenkins.remotedocker.DockerState.launchContainer(DockerState.java:285)
11:21:25 	at com.gpuopenanalytics.jenkins.remotedocker.DockerState.launchContainers(DockerState.java:199)
11:21:25 	at com.gpuopenanalytics.jenkins.remotedocker.RemoteDockerBuildWrapper.setUp(RemoteDockerBuildWrapper.java:175)
11:21:25 	at hudson.model.Build$BuildExecution.doRun(Build.java:157)
11:21:25 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:513)
11:21:25 	at hudson.model.Run.execute(Run.java:1907)
11:21:25 	at hudson.matrix.MatrixRun.run(MatrixRun.java:153)
11:21:25 	at hudson.model.ResourceController.execute(ResourceController.java:97)
11:21:25 	at hudson.model.Executor.run(Executor.java:429)
11:21:25 Recording test results
```

File: cpp/src/rolling/lead_lag_nested_detail.cuh [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    // Note: grouped_*rolling_window() trims preceding/following to
+    // the beginning/end of the group. `rolling_window()` does not.
+    // Must trim _preceding[i] so as not to go past the column start.
+    auto preceding = min(_preceding[i], i + 1);
+    return (_row_offset > (preceding - 1)) ? NULL_INDEX : (i - _row_offset);
+  }
+
+ private:
+  size_type _input_size;
+  size_type NULL_INDEX;
```
Comment (ttnghia) [inline cpp/src/rolling/lead_lag_nested_detail.cuh:72]: I don't like using `ALL_CAP` for variable name :(

In addition, for all variables that will not be changed after initialization, then please add `const` qualifier.

File: cpp/src/rolling/lead_lag_nested_detail.cuh [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+struct is_null_index_predicate_impl {
+  is_null_index_predicate_impl(size_type input_size, GatherMapIter gather_)
+    : NULL_INDEX{input_size + 1}, gather{gather_}
+  {
+  }
+
+  bool __device__ operator()(size_type i) const { return gather[i] == NULL_INDEX; }
+
+ private:
+  const size_type NULL_INDEX;
```
Comment (ttnghia) [inline cpp/src/rolling/lead_lag_nested_detail.cuh:91]:
```suggestion
  size_type const NULL_INDEX;
```

File: cpp/src/rolling/lead_lag_nested_detail.cuh [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                      input.size(), offset, preceding, following});
+
+  auto output_with_nulls = cudf::gather(table_view{std::vector<column_view>{input}},
+                                        gather_map_column->view(),
+                                        out_of_bounds_policy::NULLIFY);
+
+  if (default_outputs.is_empty()) { return std::move(output_with_nulls->release()[0]); }
+
+  // Must scatter defaults.
+  auto NULL_INDEX = size_type{input.size() + 1};
```
Comment (ttnghia) [inline cpp/src/rolling/lead_lag_nested_detail.cuh:188]: Again, `ALL_ CAP` name.

File: cpp/src/rolling/lead_lag_nested_detail.cuh [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+  // Scatter defaults into locations where LEAD/LAG computed nulls.
+  auto scattered_results = cudf::detail::scatter(
+    table_view{std::vector<column_view>{gathered_defaults->release()[0]->view()}},
+    scatter_map.begin(),
+    scatter_map_end,
+    table_view{std::vector<column_view>{output_with_nulls->release()[0]->view()}},
+    false,
+    stream);
+  return std::move(scattered_results->release()[0]);
```
Comment (ttnghia) [inline cpp/src/rolling/lead_lag_nested_detail.cuh:219]: I don't see the `mr` parameter is used for generating the result here.

**Review by ttnghia — CHANGES_REQUESTED** [review]

File: cpp/src/rolling/lead_lag_nested_detail.cuh (hunk 4 above)
Comment (mythrocks) [reply cpp/src/rolling/lead_lag_nested_detail.cuh:219]: Good catch. I'm also not using the `stream` in the first `gather()` call. I'll post a fix shortly.

File: cpp/src/rolling/lead_lag_nested_detail.cuh [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+  {
+    // Note: grouped_*rolling_window() trims preceding/following to
+    // the beginning/end of the group. `rolling_window()` does not.
+    // Must trim _preceding[i] so as not to go past the column start.
+    auto preceding = min(_preceding[i], i + 1);
+    return (_row_offset > (preceding - 1)) ? _null_index : (i - _row_offset);
+  }
+
+ private:
+  size_type const _input_size;   // Number of rows in input to LEAD/LAG.
```
Comment (ttnghia) [inline cpp/src/rolling/lead_lag_nested_detail.cuh:71]: Last time I missed this: the member variables of structs should not be private and should not have underscore `_` prefix. If you want to have public/private members then please use `class`.

File: cpp/src/rolling/lead_lag_nested_detail.cuh [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+struct is_null_index_predicate_impl {
+  is_null_index_predicate_impl(size_type input_size, GatherMapIter gather_)
+    : null_index{input_size + 1}, gather{gather_}
+  {
+  }
+
+  bool __device__ operator()(size_type i) const { return gather[i] == null_index; }
+
+ private:
+  size_type const null_index;  // Index value to use to output NULL for LEAD/LAG calculation.
```
Comment (ttnghia) [inline cpp/src/rolling/lead_lag_nested_detail.cuh:91]: There is some inconsistency here. The class above has member variables with underscore `_` prefix.

**Review by ttnghia — CHANGES_REQUESTED** [review]

**Review by ttnghia — APPROVED** [review]

**Comment by ttnghia** [comment]
Rerun tests.

**Review by rgsl888prabhu — APPROVED** [review]

File: cpp/src/rolling/lead_lag_nested_detail.cuh [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+  //
+  //    Similarly, LAG(2) is implemented using the following gather_map:
+  //        {NULL_INDEX, NULL_INDEX, 0, 1, 2...}
+  //
+  // 2. Gather input column based on the gather_map.
+  // 3. If default outputs are available, scatter contents of default_outputs`
+  //    to all positions where nulls where gathered in step 2.
+  //
+  // Note: Step 3 can be switched to use `copy_if_else()`, once it supports
+  //       nested types.
```
Comment (karthikeyann) [inline cpp/src/rolling/lead_lag_nested_detail.cuh:168]: can the entire algorithm use `copy_if_else(column_view(offset))` instead?

File: cpp/src/rolling/lead_lag_nested_detail.cuh [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+  //
+  //    For LEAD(2), the following gather_map is used:
+  //        {3, 4, 5, 6, ..., N-1, NULL_INDEX, NULL_INDEX}
+  //    where `NULL_INDEX` selects `NULL` for the gather.
+  //
+  //    Similarly, LAG(2) is implemented using the following gather_map:
+  //        {NULL_INDEX, NULL_INDEX, 0, 1, 2...}
+  //
+  // 2. Gather input column based on the gather_map.
+  // 3. If default outputs are available, scatter contents of default_outputs`
```
Comment (karthikeyann) [inline cpp/src/rolling/lead_lag_nested_detail.cuh:164]:
```suggestion
  // 3. If default outputs are available, scatter contents of `default_outputs`
```

**Review by karthikeyann — CHANGES_REQUESTED** [review]

File: cpp/src/rolling/rolling_detail.cuh [hunk 9]
```diff
@@ -22,16 +22,20 @@
 #include <cudf/column/column_device_view.cuh>
 #include <cudf/column/column_factories.hpp>
 #include <cudf/column/column_view.hpp>
+#include <cudf/concatenate.hpp>
 #include <cudf/copying.hpp>
 #include <cudf/detail/aggregation/aggregation.cuh>
 #include <cudf/detail/aggregation/aggregation.hpp>
 #include <cudf/detail/copy.hpp>
+#include <cudf/detail/gather.cuh>
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply cpp/src/rolling/rolling_detail.cuh:30]: I think you're right here. I'll remove this shortly.

File: cpp/src/rolling/rolling_detail.cuh [hunk 10]
```diff
@@ -22,16 +22,20 @@
...
 #include <cudf/detail/gather.hpp>
 #include <cudf/detail/get_value.cuh>
 #include <cudf/detail/groupby/sort_helper.hpp>
 #include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/scatter.cuh>
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply cpp/src/rolling/rolling_detail.cuh:35]: I think I left these behind when I moved the implementation to `lead_lag_nested_detail.cuh`. I will remove these shortly.

File: cpp/src/rolling/lead_lag_nested_detail.cuh [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+ */
+template <aggregation::Kind op, typename PrecedingIterator, typename FollowingIterator>
+class lead_lag_gather_map_builder {
+ public:
+  lead_lag_gather_map_builder(size_type input_size,
+                              size_type row_offset,
+                              PrecedingIterator preceding,
+                              FollowingIterator following)
+    : _input_size{input_size},
+      _null_index{input_size + 1},  // Out of input range. Gather returns null.
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply cpp/src/rolling/lead_lag_nested_detail.cuh:44]: No, this isn't needed. (This is a case of "making doubly sure", much like saving a file twice. :D) I'll retest with `size`.

File: cpp/src/rolling/lead_lag_nested_detail.cuh (hunk 7 above)
Comment (mythrocks) [reply cpp/src/rolling/lead_lag_nested_detail.cuh:168]: Sadly, no. I'm afraid not. At the moment, `copy_if_else()` doesn't handle nested types.

Your idea is sound, though. I'll be addressing the addition of nested-types to `copy_if_else()` in #8135.

**Review by karthikeyann — APPROVED** [review]

**Comment by mythrocks** [comment]
Thanks for the reviews, chaps! I'll wait for CI to pass before checking this in.

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**